helpers.go

```diff
@@ -17,48 +16,85 @@ func newStreamID() string {
 	return "STR_" + shortuuid.New()[:12]
 }
 
-func getRPCChannel(serviceName, rpc string, topic []string) string {
-	if len(topic) > 0 {
-		return fmt.Sprintf("%s|%s|%s|REQ", serviceName, rpc, strings.Join(topic, "|"))
-	} else {
-		return fmt.Sprintf("%s|%s|REQ", serviceName, rpc)
+const lowerhex = "0123456789abcdef"
+
+var channelChar = &unicode.RangeTable{
+	R16: []unicode.Range16{
+		{0x0030, 0x0039, 1}, // 0-9
+		{0x0041, 0x005a, 1}, // A-Z
+		{0x005f, 0x005f, 1}, // _
+		{0x0061, 0x007a, 1}, // a-z
+	},
+	LatinOffset: 5,
+}
+
+func appendEscaped(buf []byte, s string) []byte {
+	for _, r := range s {
+		if unicode.Is(channelChar, r) {
+			buf = append(buf, byte(r))
+		} else if r < 0x10000 {
+			buf = append(buf, `u+`...)
+			for s := 12; s >= 0; s -= 4 {
+				buf = append(buf, lowerhex[r>>uint(s)&0xF])
+			}
+		} else {
+			buf = append(buf, `u+`...)
+			for s := 28; s >= 0; s -= 4 {
+				buf = append(buf, lowerhex[r>>uint(s)&0xF])
+			}
+		}
 	}
+	return buf
 }
 
-func getHandlerKey(rpc string, topic []string) string {
-	if len(topic) > 0 {
-		return fmt.Sprintf("%s|%s", rpc, strings.Join(topic, "|"))
-	} else {
-		return rpc
+func formatChannel(prefix0, prefix1 string, topics []string, suffix string) string {
```

Comment:
we can just treat clientID as a topic instead of prefix1 here

Reply:
same with suffix

Reply:
actually it could just be `formatChannel(parts ...string)`

Reply:
resolved out of band